File: packages/backend/src/websocket.ts
import { WebSocketServer, WebSocket } from 'ws';
import { Server } from 'http';
import { v4 as uuidv4 } from 'uuid';
import { WorkflowDB } from './database';
import { WorkflowEngine } from './engine/WorkflowEngine';
import { ChatMessage } from '@workflow/shared';

interface ChatSession {
  id: string;
  workflowId: string;
  ws: WebSocket;
}

export class ChatWebSocketServer {
  private wss: WebSocketServer;
  private sessions: Map<string, ChatSession> = new Map();
  private engine: WorkflowEngine;

  constructor(server: Server) {
    this.wss = new WebSocketServer({ server, path: '/ws/chat' });
    this.engine = new WorkflowEngine();
    this.setupWebSocket();
  }

  private setupWebSocket() {
    this.wss.on('connection', (ws: WebSocket) => {
      console.log('New WebSocket connection');

      ws.on('message', async (data: Buffer) => {
        try {
          const message = JSON.parse(data.toString());
          await this.handleMessage(ws, message);
        } catch (error: any) {
          console.error('WebSocket message error:', error);
          ws.send(JSON.stringify({
            type: 'error',
            error: error.message
          }));
        }
      });

      ws.on('close', () => {
        // Clean up session
        for (const [sessionId, session] of this.sessions.entries()) {
          if (session.ws === ws) {
            this.sessions.delete(sessionId);
            console.log(`Session ${sessionId} closed`);
            break;
          }
        }
      });
    });
  }

  private async handleMessage(ws: WebSocket, message: any) {
    switch (message.type) {
      case 'start_session':
        await this.startSession(ws, message.workflowId);
        break;

      case 'chat_message':
        await this.handleChatMessage(ws, message.sessionId, message.content);
        break;

      default:
        ws.send(JSON.stringify({
          type: 'error',
          error: `Unknown message type: ${message.type}`
        }));
    }
  }

  private async startSession(ws: WebSocket, workflowId: string) {
    const workflow = WorkflowDB.getWorkflow(workflowId);
    if (!workflow) {
      ws.send(JSON.stringify({
        type: 'error',
        error: 'Workflow not found'
      }));
      return;
    }

    const sessionId = uuidv4();
    WorkflowDB.createChatSession(sessionId, workflowId);

    this.sessions.set(sessionId, {
      id: sessionId,
      workflowId,
      ws
    });

    ws.send(JSON.stringify({
      type: 'session_started',
      sessionId,
      workflowId
    }));

    console.log(`Started chat session ${sessionId} for workflow ${workflowId}`);
  }

  private async handleChatMessage(ws: WebSocket, sessionId: string, content: string) {
    const session = this.sessions.get(sessionId);
    if (!session) {
      ws.send(JSON.stringify({
        type: 'error',
        error: 'Session not found'
      }));
      return;
    }

    // Save user message
    const userMessage: ChatMessage = {
      id: uuidv4(),
      sessionId,
      role: 'user',
      content,
      timestamp: new Date().toISOString()
    };
    WorkflowDB.addChatMessage(userMessage);

    // Send acknowledgment
    ws.send(JSON.stringify({
      type: 'message_received',
      message: userMessage
    }));

    try {
      // Execute workflow with chat message as trigger
      const execution = await this.engine.executeFromTrigger(session.workflowId, {
        message: content,
        sessionId
      });

      // Get response from workflow execution
      const responseNode = Object.keys(execution.context).find(key =>
        key !== 'trigger' && execution.context[key]?.message
      );

      let responseContent = 'Workflow executed successfully';
      if (responseNode && execution.context[responseNode]?.message) {
        responseContent = execution.context[responseNode].message;
      } else if (execution.context.response?.message) {
        responseContent = execution.context.response.message;
      } else {
        // Check for AI agent response
        const aiAgentNode = Object.keys(execution.context).find(key =>
          execution.context[key]?.text
        );
        if (aiAgentNode) {
          responseContent = execution.context[aiAgentNode].text;
        }
      }

      // Save assistant message
      const assistantMessage: ChatMessage = {
        id: uuidv4(),
        sessionId,
        role: 'assistant',
        content: responseContent,
        timestamp: new Date().toISOString()
      };
      WorkflowDB.addChatMessage(assistantMessage);

      // Send response
      ws.send(JSON.stringify({
        type: 'chat_response',
        message: assistantMessage,
        executionId: execution.id
      }));

    } catch (error: any) {
      console.error('Workflow execution error:', error);

      const errorMessage: ChatMessage = {
        id: uuidv4(),
        sessionId,
        role: 'assistant',
        content: `Error: ${error.message}`,
        timestamp: new Date().toISOString()
      };
      WorkflowDB.addChatMessage(errorMessage);

      ws.send(JSON.stringify({
        type: 'chat_response',
        message: errorMessage,
        error: error.message
      }));
    }
  }

  getActiveSessions(): number {
    return this.sessions.size;
  }
}

File: backend/websocket/websocket.go
package websocket

import (
	"encoding/json"
	"log"
	"sync"
	"time"

	"github.com/google/uuid"
	"github.com/gorilla/websocket"
	"github.com/workflow-orchestration/backend/database"
	"github.com/workflow-orchestration/backend/engine"
	"github.com/workflow-orchestration/backend/models"
)

type ChatSession struct {
	ID         string
	WorkflowID string
	Conn       *websocket.Conn
}

type ChatWebSocketServer struct {
	db       *database.DB
	engine   *engine.WorkflowEngine
	sessions map[string]*ChatSession
	mu       sync.RWMutex
}

func NewChatWebSocketServer(db *database.DB, engine *engine.WorkflowEngine) *ChatWebSocketServer {
	return &ChatWebSocketServer{
		db:       db,
		engine:   engine,
		sessions: make(map[string]*ChatSession),
	}
}

func (cws *ChatWebSocketServer) HandleConnection(conn *websocket.Conn) {
	defer conn.Close()

	for {
		_, message, err := conn.ReadMessage()
		if err != nil {
			log.Printf("WebSocket read error: %v", err)
			// Clean up session
			cws.cleanupConnection(conn)
			break
		}

		var msg map[string]interface{}
		if err := json.Unmarshal(message, &msg); err != nil {
			log.Printf("JSON unmarshal error: %v", err)
			continue
		}

		if err := cws.handleMessage(conn, msg); err != nil {
			log.Printf("Handle message error: %v", err)
			cws.sendError(conn, err.Error())
		}
	}
}

func (cws *ChatWebSocketServer) handleMessage(conn *websocket.Conn, msg map[string]interface{}) error {
	msgType, ok := msg["type"].(string)
	if !ok {
		return nil
	}

	switch msgType {
	case "start_session":
		return cws.startSession(conn, msg)
	case "chat_message":
		return cws.handleChatMessage(conn, msg)
	}

	return nil
}

func (cws *ChatWebSocketServer) startSession(conn *websocket.Conn, msg map[string]interface{}) error {
	workflowID, ok := msg["workflowId"].(string)
	if !ok {
		return cws.sendError(conn, "workflowId is required")
	}

	// Verify workflow exists
	workflow, err := cws.db.GetWorkflow(workflowID)
	if err != nil {
		return err
	}
	if workflow == nil {
		return cws.sendError(conn, "Workflow not found")
	}

	sessionID := uuid.New().String()
	if err := cws.db.CreateChatSession(sessionID, workflowID); err != nil {
		return err
	}

	session := &ChatSession{
		ID:         sessionID,
		WorkflowID: workflowID,
		Conn:       conn,
	}

	cws.mu.Lock()
	cws.sessions[sessionID] = session
	cws.mu.Unlock()

	return cws.send(conn, map[string]interface{}{
		"type":       "session_started",
		"sessionId":  sessionID,
		"workflowId": workflowID,
	})
}

func (cws *ChatWebSocketServer) handleChatMessage(conn *websocket.Conn, msg map[string]interface{}) error {
	sessionID, ok := msg["sessionId"].(string)
	if !ok {
		return cws.sendError(conn, "sessionId is required")
	}

	content, ok := msg["content"].(string)
	if !ok {
		return cws.sendError(conn, "content is required")
	}

	cws.mu.RLock()
	session, exists := cws.sessions[sessionID]
	cws.mu.RUnlock()

	if !exists {
		return cws.sendError(conn, "Session not found")
	}

	// Save user message
	userMessage := &models.ChatMessage{
		ID:        uuid.New().String(),
		SessionID: sessionID,
		Role:      "user",
		Content:   content,
		Timestamp: time.Now(),
	}
	if err := cws.db.AddChatMessage(userMessage); err != nil {
		return err
	}

	// Send acknowledgment
	if err := cws.send(conn, map[string]interface{}{
		"type":    "message_received",
		"message": userMessage,
	}); err != nil {
		return err
	}

	// Execute workflow
	execution, err := cws.engine.ExecuteFromTrigger(session.WorkflowID, map[string]interface{}{
		"message":   content,
		"sessionId": sessionID,
	})

	if err != nil {
		log.Printf("Workflow execution error: %v", err)
		errorMessage := &models.ChatMessage{
			ID:        uuid.New().String(),
			SessionID: sessionID,
			Role:      "assistant",
			Content:   "Error: " + err.Error(),
			Timestamp: time.Now(),
		}
		cws.db.AddChatMessage(errorMessage)

		return cws.send(conn, map[string]interface{}{
			"type":    "chat_response",
			"message": errorMessage,
			"error":   err.Error(),
		})
	}

	// Extract response from execution context
	responseContent := "Workflow executed successfully"

	// Check for response node output
	if responseData, ok := execution.Context["response"].(map[string]interface{}); ok {
		if msg, ok := responseData["message"].(string); ok {
			responseContent = msg
		}
	} else {
		// Look for AI agent response
		for key, value := range execution.Context {
			if key != "trigger" {
				if data, ok := value.(map[string]interface{}); ok {
					if text, ok := data["text"].(string); ok && text != "" {
						responseContent = text
						break
					}
					if msg, ok := data["message"].(string); ok && msg != "" {
						responseContent = msg
						break
					}
				}
			}
		}
	}

	// Save assistant message
	assistantMessage := &models.ChatMessage{
		ID:        uuid.New().String(),
		SessionID: sessionID,
		Role:      "assistant",
		Content:   responseContent,
		Timestamp: time.Now(),
	}
	if err := cws.db.AddChatMessage(assistantMessage); err != nil {
		return err
	}

	// Send response
	return cws.send(conn, map[string]interface{}{
		"type":        "chat_response",
		"message":     assistantMessage,
		"executionId": execution.ID,
	})
}

func (cws *ChatWebSocketServer) send(conn *websocket.Conn, data interface{}) error {
	return conn.WriteJSON(data)
}

func (cws *ChatWebSocketServer) sendError(conn *websocket.Conn, errorMsg string) error {
	return cws.send(conn, map[string]interface{}{
		"type":  "error",
		"error": errorMsg,
	})
}

func (cws *ChatWebSocketServer) cleanupConnection(conn *websocket.Conn) {
	cws.mu.Lock()
	defer cws.mu.Unlock()

	for sessionID, session := range cws.sessions {
		if session.Conn == conn {
			delete(cws.sessions, sessionID)
			log.Printf("Session %s closed", sessionID)
			break
		}
	}
}

func (cws *ChatWebSocketServer) GetActiveSessions() int {
	cws.mu.RLock()
	defer cws.mu.RUnlock()
	return len(cws.sessions)
}
